neqo-transport/src/cc.rs

```diff
@@ -117,31 +119,40 @@ impl CongestionControl {
                 continue;
             }
 
-            if self.congestion_window < self.ssthresh {
-                self.congestion_window += pkt.size;
-                qinfo!([self], "slow start");
-                qlog::congestion_state_updated(
-                    &mut self.qlog,
-                    &mut self.qlog_curr_cong_state,
-                    CongestionState::SlowStart,
-                );
-            } else {
-                self.congestion_window += (MAX_DATAGRAM_SIZE * pkt.size) / self.congestion_window;
-                qinfo!([self], "congestion avoidance");
-                qlog::congestion_state_updated(
-                    &mut self.qlog,
-                    &mut self.qlog_curr_cong_state,
-                    CongestionState::CongestionAvoidance,
-                );
-            }
-            qlog::metrics_updated(
+            self.acked_bytes += pkt.size;
+        }
+        qtrace!([self], "ACK received, acked_bytes = {}", self.acked_bytes);
+
+        // Slow start, up to the slow start threshold.
+        if self.congestion_window < self.ssthresh {
+            let increase = min(self.ssthresh - self.congestion_window, self.acked_bytes);
+            self.congestion_window += increase;
+            self.acked_bytes -= increase;
+            qinfo!([self], "slow start += {}", increase);
+            qlog::congestion_state_updated(
                 &mut self.qlog,
-                &[
-                    QlogMetric::CongestionWindow(self.congestion_window),
-                    QlogMetric::BytesInFlight(self.bytes_in_flight),
-                ],
+                &mut self.qlog_curr_cong_state,
+                CongestionState::SlowStart,
+            );
+        }
+        // Congestion avoidance, above the slow start threshold.
+        if self.acked_bytes >= self.congestion_window {
+            self.congestion_window += MAX_DATAGRAM_SIZE;
+            self.acked_bytes -= MAX_DATAGRAM_SIZE;
```

Comment:
acked_bytes are increase on every packet by the size of the packet. These lines mean hat we are going to double cwnd on each RTT.
I think this should be set to 0.

Reply:
That's a good catch.  This should be self.acked_bytes -= congestion_window (so we keep any overage) and should be reordered with the line above.

Reply:
I've added a test that catches the bug (and fixed the bug).